### PR TITLE
Add transfer learning episode

### LINKED ISSUE #349: Transfer learning episode?
I see there's an old closed issue related to this topic — #46. I would strongly argue in favor of adding transfer learning to these materials. Transfer learning is extremely common in research applications of deep learning because many domains lack sufficiently large datasets. Happy to try to get something started in a new (perhaps optional) episode or as an extension of "advanced layer types".

### PR DESCRIPTION
**Pull request by svenvanderburg**
Fixes #349

### CODE REVIEW

File: episodes/5-transfer-learning.Rmd [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+::: questions
+- How do I apply a pre-trained model to my data?
+:::
+
+::: objectives
+- Apply a state-of-the-art pre-trained network on your own dataset
+:::
+
+## What is transfer learning?
+Instead of training a model from scratch, with transfer learning you make use of models that are trained on another machine learning task. The pre-trained network captures generic knowledge about the domain during pre-training and will only be 'fine-tuned' to the specifics of your dataset.
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:16]: I would remove the 'domain' reference for its ambiguity; it could refer to both the generic domain but also to the domain of your specific dataset.

```suggestion
Instead of training a model from scratch, with transfer learning you make use of models that are trained on another machine learning task. The pre-trained network captures generic knowledge during pre-training and will only be 'fine-tuned' to the specifics of your dataset.
```

File: episodes/5-transfer-learning.Rmd [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+teaching: 20
+exercises: 30
+---
+
+::: questions
+- How do I apply a pre-trained model to my data?
+:::
+
+::: objectives
+- Apply a state-of-the-art pre-trained network on your own dataset
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:12]:
```suggestion
- Adapt a state-of-the-art pre-trained network to your own dataset
```

File: episodes/5-transfer-learning.Rmd [hunk 3]
```diff
@@ -0,0 +1,235 @@
+---
+title: "Transfer learning"
+teaching: 20
+exercises: 30
+---
+
+::: questions
+- How do I apply a pre-trained model to my data?
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:8]:
```suggestion
- How do I use a pre-trained model to my data?
```

File: episodes/5-transfer-learning.Rmd [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+::: objectives
+- Apply a state-of-the-art pre-trained network on your own dataset
+:::
+
+## What is transfer learning?
+Instead of training a model from scratch, with transfer learning you make use of models that are trained on another machine learning task. The pre-trained network captures generic knowledge about the domain during pre-training and will only be 'fine-tuned' to the specifics of your dataset.
+
+An example: Let's say that you want to train a model to classify images of different dog breeds. You could make use of a pre-trained network that learned how to classify images of dogs and cats. The pre-trained network will not know anything about different dog breeds, but it will have captured some general knowledge of, on a high-level, what dogs look like, and on a low-level all the different features (eyes, ears, paws, fur) that make up an image of a dog. Further training this model on your dog breed dataset is a much easier task than training from scratch, because the model can use the general knowledge captured in the pre-trained network.
+
+In this episode we will learn how use Keras to apply a state-of-the-art pre-trained model to the dollarstreet-10 dataset.
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:20]:
```suggestion
In this episode we will learn how use Keras to adapt a state-of-the-art pre-trained model to the dollarstreet-10 dataset.
```

Comment (carschno) [inline episodes/5-transfer-learning.Rmd:20]: I couldn't find a reference to the `-10` anywhere, have I missed it? Otherwise, I would propose to stick with the exact spelling to prevent confusions, including word splitting and capitalisation.

Also, how about adding a link to the dataset?

```suggestion
In this episode we will learn how use Keras to apply a state-of-the-art pre-trained model to the [Dollar Street Dataset](https://mlcommons.org/datasets/dollar-street/).
```

File: episodes/5-transfer-learning.Rmd [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+
+
+## 1. Formulate / Outline the problem and 2. Identify inputs and outputs
+
+
+Just like in the previous episode, we use the dollar street 10 dataset.
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:26]:
```suggestion
Just like in the previous episode, we use the Dollar Street Dataset. 
```

File: episodes/5-transfer-learning.Rmd [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+trained on images of 160 x 160 pixels. To deal with this, we add an upscale layer
+that resizes the images to 160 x 160 pixels during training and prediction.
+```python
+# upscale layer
+method = tf.image.ResizeMethod.BILINEAR
+upscale = keras.layers.Lambda(
+  lambda x: tf.image.resize_with_pad(x, 160, 160, method=method))(inputs)
+```
+
+From the `keras.applications` module we use the `DenseNet121` architecture.
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:65]:
```suggestion
We use the `DenseNet121` architecture from the `keras.applications` module. 
```

File: episodes/5-transfer-learning.Rmd [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+method = tf.image.ResizeMethod.BILINEAR
+upscale = keras.layers.Lambda(
+  lambda x: tf.image.resize_with_pad(x, 160, 160, method=method))(inputs)
+```
+
+From the `keras.applications` module we use the `DenseNet121` architecture. 
+This architecture was proposed by the paper: [Densely Connected Convolutional Networks (CVPR 2017)](https://arxiv.org/abs/1608.06993). It is trained on the [Imagenet](https://www.image-net.org/) dataset, which contains 14,197,122 annotated images according to the WordNet hierarchy with over 20,000 classes.
+
+We will have a look at the architecture later, for now it is enough to know
+that it is a convolutional neural network with 121 layers that was designed
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:69]: I wonder: is it the large amount of data or the large number of parameter that leads to the improved accuracy? Or: is it really the transfer learning aspect that gives the boost?
My guess is that it is both, but I think that we should have a more specific answer.

File: episodes/5-transfer-learning.Rmd [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+```python
+out = base_model.output
+out = keras.layers.Flatten()(out)
+out = keras.layers.BatchNormalization()(out)
+out = keras.layers.Dense(128, activation='relu')(out)
+out = keras.layers.Dropout(0.5)(out)
+out = keras.layers.BatchNormalization()(out)
+out = keras.layers.Dense(64, activation='relu')(out)
+out = keras.layers.Dropout(0.5)(out)
+out = keras.layers.Dense(10)(out)
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:114]: Not really important, but the same hyper-parameters (layers, layer sizes, dropout rates) as in the previous section for consistency might prevent confusion here.

File: episodes/5-transfer-learning.Rmd [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+```python
+model.compile(optimizer='adam',
+              loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+              metrics=['accuracy'])
+```
+
+### 2. Train the model
+Define the early stopper:
+```python
+keras.callbacks.EarlyStopping(monitor='val_accuracy',
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:188]: I understand you need to store the `EarlyStopping` object in a variable to use it in the next step.

```suggestion
early_stopper = keras.callbacks.EarlyStopping(monitor='val_accuracy',
```

File: episodes/5-transfer-learning.Rmd [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+### 4. (Optional) Try out other pre-trained neural networks
+Train and evaluate another pre-trained model from https://keras.io/api/applications/. How does it compare to DenseNet121?
+
+
+::::
+:::
+
+## Concluding: The power of transfer learning
+In many domains large networks pre-trained on vast amounts of data are available, such as in computer vision and natural language processing. Using transfer learning you can benefit from the knowledge stored in a pre-trained network that was captured from another machine learning task. In many fields transfer learning will outperform models trained from scratch, especially if your dataset is small or of poor quality.
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:230]: Minor stylistic changes to shorten the sentences a bit:

```suggestion
In many domains, large networks are available that have been trained on vast amounts of data, such as in computer vision and natural language processing. Using transfer learning, you can benefit from the knowledge that was captured from another machine learning task. In many fields, transfer learning will outperform models trained from scratch, especially if your dataset is small or of poor quality.
```

File: episodes/5-transfer-learning.Rmd [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+
+::: keypoints
+- Large pre-trained capture generic knowledge about a domain
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:233]:
```suggestion
- Large pre-trained models capture generic knowledge about a domain
```

File: episodes/5-transfer-learning.Rmd [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+- Use the `keras.applications` module to easily use pre-trained models on your own datasets
```
Comment (carschno) [inline episodes/5-transfer-learning.Rmd:234]:
```suggestion
- Use the `keras.applications` module to easily use pre-trained models for your own datasets
```

**Review by carschno — COMMENTED** [review]
Conceptually, I think we need to be able to answer the question: is the improvement due to the additional data, or to the larger model (see comment in-line).

Otherwise, some minor suggestions.

File: episodes/5-transfer-learning.Rmd (hunk 1 above)
Comment (svenvanderburg) [reply episodes/5-transfer-learning.Rmd:16]: Good point! Totally agree.

File: episodes/5-transfer-learning.Rmd (hunk 4 above)
Comment (svenvanderburg) [reply episodes/5-transfer-learning.Rmd]: Good point! You couldn't find it, because it is a dataset that I made myself. But I will call it Dollar Street Dataset, but link to https://zenodo.org/records/10970014 instead of the mlcommons dataset from which it is derived.

File: episodes/5-transfer-learning.Rmd (hunk 9 above)
Comment (svenvanderburg) [reply episodes/5-transfer-learning.Rmd:188]: Whoops!

File: episodes/5-transfer-learning.Rmd (hunk 8 above)
Comment (svenvanderburg) [reply episodes/5-transfer-learning.Rmd:114]: Good point, also it is overly complex.

I now use:
```python
out = base_model.output
out = keras.layers.Flatten()(out)
out = keras.layers.BatchNormalization()(out)
out = keras.layers.Dense(50, activation='relu')(out)
out = keras.layers.Dropout(0.5)(out)
out = keras.layers.Dense(10)(out)
```
